### PR TITLE
Fixed infinite save loop #16395

### PR DESCRIPTION
**Pull request by rudiservo**
Hello!

* Type: bug fix
* Link to issue: [#16395]

**In raising this pull request, I confirm the following:**

- [X] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)
- [X] I have checked that another pull request for this purpose does not exist
- [X] I wrote some tests for this PR
- [X] I have updated the relevant CHANGELOG
- [ ] I have created a PR for the [documentation](https://github.com/phalcon/docs) about this change

Small description of change:
Added an array with already visited objects by the save or the SaveRelated chain functions in order to avoid save loops
Performance Tests where made between SplObjectStorage and the PHP Array.
There are two function to get a unique ID from a Php object, spl_object_hash and spl_object_id.
Using spl_object_hash, SplObjectStorage is faster and uses less memory, spl_object_storage is the function SplObjectStorage uses.
Using the spl_object_id the php Array is faster and uses less memory by a small margin, however since Zephir cannot pass by reference, SplObjectStorage is the better option.

Thanks

### CODE REVIEW

**Review by niden — APPROVED** [review]

File: phalcon/Mvc/Model.zep [hunk 1]
```diff
@@ -2695,7 +2721,7 @@ abstract class Model extends AbstractInjectionAware implements EntityInterface,
         /**
          * Change the dirty state to persistent
          */
-        if success {
+        if true === success{
```
Comment (Jeckerson) [inline phalcon/Mvc/Model.zep:2724]:
```suggestion
        if true === success {
```

File: phalcon/Mvc/Model.zep [hunk 2]
```diff
@@ -4981,6 +5010,7 @@ abstract class Model extends AbstractInjectionAware implements EntityInterface,
             manager = <ManagerInterface> this->getModelsManager();
 
         for name, record in related {
+
```
Comment (Jeckerson) [inline phalcon/Mvc/Model.zep:5013]:
```suggestion
```

File: tests/database/Mvc/Model/SaveCest.php [hunk 3]
```diff
@@ -644,6 +644,35 @@ public function mvcModelSaveWithRelatedManyAndBelongsRecordsProperty(DatabaseTes
...
+        $invoicesMigration = new InvoicesMigration($connection);
+        $invoicesMigration->insert(77, 1, 0, uniqid('inv-', true));
+
+        $customersMigration = new CustomersMigration($connection);
+        $customersMigration->insert(1, 1, 'test_firstName_1', 'test_lastName_1');
+
+        $customer = Customers::findFirst(1);
+        $invoice = Invoices::findFirst(77);
+        $invoice->customer = $customer;
+        $customer->invoices = [ $invoice ];
```
Comment (Jeckerson) [inline tests/database/Mvc/Model/SaveCest.php:672]:
```suggestion
        $customer->invoices = [$invoice];
```

File: tests/database/Mvc/Model/SaveCest.php [hunk 4]
```diff
@@ -644,6 +644,35 @@ public function mvcModelSaveWithRelatedManyAndBelongsRecordsProperty(DatabaseTes
         $I->assertEquals($expected, $actual);
     }
 
+    /**
+     * Tests Phalcon\Mvc\Model\ :: save() Infinite Loop
+     *
+     * @author Phalcon Team <team@phalcon.io>
+     * @since  2023-08-09
+     * @issue  #16395
```
Comment (Jeckerson) [inline tests/database/Mvc/Model/SaveCest.php:652]:
```suggestion
     * @issue  https://github.com/phalcon/cphalcon/issues/16395
```

File: phalcon/Mvc/Model.zep [hunk 5]
```diff
@@ -2579,12 +2581,36 @@ abstract class Model extends AbstractInjectionAware implements EntityInterface,
...
+    }
+
+    /**
+     * Inserted or updates model instance, expects a visited list of objects.
+     *
+     * @param CollectionInterface $visited
+     *
+     * @return bool
+     */
+
```
Comment (Jeckerson) [inline phalcon/Mvc/Model.zep:2599]:
```suggestion
```

Comment (rudiservo) [reply phalcon/Mvc/Model.zep]: done

**Comment by Jeckerson** [comment]
@rudiservo please rebase this branch.

**Pull request merged**